narwhals/_pandas_like/namespace.py

```diff
@@ -250,3 +262,97 @@ def concat(
                 backend_version=self._backend_version,
             )
         raise NotImplementedError
+
+    def when(
+        self,
+        *predicates: IntoPandasLikeExpr | Iterable[IntoPandasLikeExpr],
+        **constraints: Any,
+    ) -> PandasWhen:
+        plx = self.__class__(self._implementation, self._backend_version)
+        if predicates:
+            condition = plx.all_horizontal(*flatten(predicates))
+        elif constraints:
+            condition = plx.all_horizontal(
+                *flatten([plx.col(key) == value for key, value in constraints.items()])
+            )
+        else:
+            msg = "at least one predicate or constraint must be provided"
+            raise TypeError(msg)
```

Comment:
what if both `predicates` and `constraints` are passed?

I'd suggest we leave `constraints` out from the initial implementation, we don't support it in `filter` either

Reply:
Sounds good, makes things much simpler

I believe in polars as well passing both is disallowed, can check as am not sure

Reply:
@MarcoGorelli

It should be removed now